src/Abp.Zero.EntityFramework/EntityHistory/EntityHistoryHelper.cs

```diff
@@ -487,54 +475,94 @@ private void UpdateChangeSet(DbContext context, EntityChangeSet changeSet)
 
-                var foreignKeys = entityType.NavigationProperties;
-
-                foreach (var foreignKey in foreignKeys)
+                foreach(var propertyName in foreignKeyPropertyNames)
                 {
-                    foreach (var property in foreignKey.GetDependentProperties())
+                    var propertyEntry = entityEntry.Property(propertyName);
+                    var propertyInfo = entityEntry.GetPropertyInfo(propertyName);
+                    var propertyChange = entityChange.PropertyChanges.FirstOrDefault(pc => pc.PropertyName == propertyName);
+                    var oldValue = propertyEntry.GetOriginalValue();
+                    var newValue = propertyEntry.GetNewValue();
+
+                    if (propertyChange == null)
                     {
-                        var propertyEntry = entityEntry.Property(property.Name);
-                        var propertyChange = entityChange.PropertyChanges.FirstOrDefault(pc => pc.PropertyName == property.Name);
+                        if (propertyEntry.HasChanged())
+                        {
+                            // Add foreign key
+                            entityChange.PropertyChanges.Add(CreateEntityPropertyChange(oldValue, newValue, propertyInfo));
+                        }
+
+                        continue;
+                    }
 
-                        //make sure test case cover post saving (for foreign key update)
-                        if (propertyChange == null)
+                    if (propertyChange.OriginalValue == propertyChange.NewValue)
+                    {
+                        var newPropertyChange = CreateEntityPropertyChange(oldValue, newValue, propertyInfo);
+                        if (propertyChange.NewValue == newPropertyChange.NewValue)
+                        {
+                            // No change
+                            entityChange.PropertyChanges.Remove(propertyChange);
+                        }
+                        else
                         {
-                            if (propertyEntry.HasChanged())
-                            {
-                                var propertyInfo = entityEntry.GetPropertyInfo(property.Name);
-
-                                // Add foreign key
-                                entityChange.PropertyChanges.Add(new EntityPropertyChange
-                                {
-                                    NewValue = propertyEntry.CurrentValue.ToJsonString(),
-                                    OriginalValue = propertyEntry.OriginalValue.ToJsonString(),
-                                    PropertyName = property.Name,
-                                    PropertyTypeFullName = propertyInfo.PropertyType.FullName
-                                });
-                            }
-
-                            continue;
+                            // Update foreign key
+                            propertyChange.NewValue = newPropertyChange.NewValue;
```

Comment:
```suggestion
                            propertyChange.SetNewValue(newValue);
```

where `SetNewValue` does `TruncateWithPostfix`.

Let's also change the setters for `NewValue` and `OldValue` to `protected set`.